packages/block-editor/src/components/block-toolbar/utils.js

```diff
@@ -0,0 +1,263 @@
+} = window;
+
+/**
+ * Hook that creates a showMover state, as well as debounced show/hide callbacks
+ */
+export function useDebouncedShowMovers( {
+	ref,
+	isFocused,
+	debounceTimeout = 500,
+} ) {
+	const [ showMovers, setShowMovers ] = useState( false );
+	const timeoutRef = useRef();
+
+	const getIsHovered = () => {
+		return ref && ref.current.matches( ':hover' );
+	};
+
+	const shouldHideMovers = () => {
+		const isHovered = getIsHovered();
+
+		return ! isFocused && ! isHovered;
+	};
+
+	const debouncedShowMovers = useCallback(
+		( event ) => {
+			if ( event ) {
+				event.stopPropagation();
+			}
+
+			const timeout = timeoutRef.current;
+
+			if ( timeout && clearTimeout ) {
+				clearTimeout( timeout );
+			}
+			if ( ! showMovers ) {
+				setShowMovers( true );
+			}
+		},
+		[ showMovers ]
+	);
+
+	const debouncedHideMovers = useCallback(
+		( event ) => {
+			if ( event ) {
+				event.stopPropagation();
+			}
+
+			timeoutRef.current = setTimeout( () => {
+				if ( shouldHideMovers() ) {
+					setShowMovers( false );
+				}
+			}, debounceTimeout );
+		},
+		[ isFocused ]
+	);
+
+	return {
+		showMovers,
+		debouncedShowMovers,
+		debouncedHideMovers,
+	};
+}
+
+/**
+ * Hook that provides a showMovers state and gesture events for DOM elements
+ * that interact with the showMovers state.
+ */
+export function useShowMoversGestures( { ref, debounceTimeout = 500 } ) {
+	const [ isFocused, setIsFocused ] = useState( false );
+	const {
+		showMovers,
+		debouncedShowMovers,
+		debouncedHideMovers,
+	} = useDebouncedShowMovers( { ref, debounceTimeout, isFocused } );
+
+	const registerRef = useRef( false );
+
+	const isFocusedWithin = () => {
+		return ref && ref.current.contains( document.activeElement );
+	};
+
+	useEffect( () => {
+		const node = ref.current;
+
+		const handleOnFocus = () => {
+			if ( isFocusedWithin() ) {
+				setIsFocused( true );
+				debouncedShowMovers();
+			}
+		};
+
+		const handleOnBlur = () => {
+			if ( ! isFocusedWithin() ) {
+				setIsFocused( false );
+				debouncedHideMovers();
+			}
+		};
+
+		/**
+		 * Events are added via DOM events (vs. React synthetic events),
+		 * as the child React components swallow mouse events.
+		 */
+		if ( node && ! registerRef.current ) {
+			node.addEventListener( 'focus', handleOnFocus, true );
+			node.addEventListener( 'blur', handleOnBlur, true );
+			registerRef.current = true;
+		}
+
+		return () => {
+			if ( node ) {
+				node.removeEventListener( 'focus', handleOnFocus );
+				node.removeEventListener( 'blur', handleOnBlur );
+			}
+		};
+	}, [
+		ref,
+		registerRef,
+		setIsFocused,
+		debouncedShowMovers,
+		debouncedHideMovers,
+	] );
+
+	return {
+		showMovers,
+		gestures: {
+			onMouseMove: debouncedShowMovers,
+			onMouseLeave: debouncedHideMovers,
+		},
+	};
+}
+
+const EDITOR_SELECTOR = '.editor-styles-wrapper';
+const VIEWPORT_BREAKPOINT = 782;
+
+/**
+ * This is experimental.
+ */
+export function useExperimentalToolbarPositioning( { ref } ) {
```

Comment:
Is this used to adjust the left position of the toolbar when there's not enough space? 
If it's the case, I believe we should start without it first and try to implement this in Popover instead. Popover already adjusts the position according to the right edge, we'll have to update it to adjust according the left one as well.

In my testing, this is a bit unstable now, sometimes the toolbar is not positionned properly when resizing the canvas. I believe it's probably because there's two competing positionning mechanism that are running in two different places completely.

Reply:
>  I believe it's probably because there's two competing positionning mechanism that are running in two different places completely.

Yes, unfortunately, that's the case right now. Ideally, we'd only leverage the positioning system from Popover. However, I wasn't able to get it to account for the negative offset of the Mover/Block selector, illustrated here:

![](https://user-images.githubusercontent.com/2322354/74156598-aae54400-4be4-11ea-808c-8092f8319b87.png)

To simplify things, an alternative (for now) would be to reduce the "double" offset of the Mover + Block selector. From my experiments, this will enable us to only use the Popover positioner.

However, the Toolbar would appear flush with the content by default, like so:

![Screen Capture on 2020-02-14 at 09-50-56](https://user-images.githubusercontent.com/2322354/74541592-e434f100-4f0f-11ea-83ba-7c5272ddf203.gif)

(cc'ing @jasmussen / @mtias for thoughts!)

Reply:
Also cc @ellatrix here as she worked on the right shifting of the toolbar.